### ISSUE TITLE
Enforce import index style

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#9]
I want to enforce `require('.');` and `from '.';`. Before Node.js 4 you had to do `'./'`. In Node.js 4 you can do: `'.'`.

Should catch the following cases:
- `'./'`
- `'./index'`
- `'./index.js'`

Should be easy to implement auto-fixing for this rule.

**Comment by jamestalmage** [comment]
Is it premature? Should we enforce `./` until `0.12` is phased out?

**Comment by jfmengels** [comment]
Well, the rule could exist with an option of what is the preferred version, and could be rectivated later by default in XO.

**Comment by jamestalmage** [comment]
Yeah. That's what I was thinking.

**Comment by sindresorhus** [comment]
I don't see the point in an option. 0.12 is already being phased out. It went into maintenance mode early this month. I'd rather have a simple rule without any options that we can add to `esnext` in XO for now.

**Comment by jamestalmage** [comment]
My thought is that most of what I work on still supports `0.12`, and will need to for the next 12 months.

 `.` vs `./` isn't that big of deal, but it would be nice to call out `./index` and `./index.js`.

Either way, the rule should also handle relative references to `index.js`:

- `../../index.js` -> `../../`

**Comment by sindresorhus** [comment]
Alright. So how about we fail on `./index` and `./index.js` and allow `.` and `./`, but in a year we change it to only `.`?

**Comment by jfmengels** [comment]
If it gets in XO with `esnext` and not as the default, I don't see why it couldn't be `.` already. Or give it an option to prefer `./` or `.`, that way, we won't have to change it in a year and "migrate" the codebases that use `esnext`.

**Comment by sindresorhus** [comment]
> Or give it an option to prefer ./ or ., that way, we won't have to change it in a year and "migrate" the codebases that use esnext.

I was hoping to not need an option, but I agree with @jfmengels. That sounds like the best solution. Should default to `.`, so we can just remove the option in a year.

**Comment by jamestalmage** [comment]
You can still call `require` from transpiled code, and I don't think Babel handles converting `.` => `./`.

So simply using XO's `esnext` option is going to create some confusion I think.

IMO, the best thing to do would be to default to `./` as the requirement, and `.` should be an error. That keeps it consistent and backward compatible.

I don't see the point of trying to drop that trailing slash. Sure it's an extra character. But IMO, having an explicit backward compatible rule trumps that.

**Comment by sindresorhus** [comment]
Ok. I'm good with `./`.

We can change it to `.` later and I can use a codemod to update all my modules.

**Comment by jamestalmage** [comment]
:+1:

**Comment by ayrton** [comment]
I wrote this earlier today for closed project, happy to open a PR:

```js
module.exports = function(context) {
  return {
    'ImportDeclaration': function(node) {
      if (node.source.value.slice(-1) === '/') {
        context.report(node, 'Omit trailing slashes in your imports. Rename to `' + node.source.value.slice(0, -1) + '`');
      }

      if (node.source.value.match(/\/index\.js/)) {
        var index = node.source.value.indexOf('/index.js')
        context.report(node, 'Omit `/index.js` in your imports. Rename to `' + node.source.value.slice(0, index) + '`');
      }
    }
  }
}
```

**Comment by jfmengels** [comment]
@ayrton Great! And yes, please make a PR :)

A few remarks already though:
- We would like support for `require` too
- You don't handle the case of `./index`

Thanks for doing this!

**Comment by sindresorhus** [comment]
Also needs to handle `require()`.

**Comment by jfmengels** [comment]
@ayrton Are you still up for making a PR for this? No worries if you are not.

**Comment by ayrton** [comment]
@jfmengels I'm currently on holiday - so wouldn't mind if someone took over. Won't be able to work on it for another two weeks.

**Comment by jfmengels** [comment]
@ayrton Sure, no problem. Maybe I'll tackle this, maybe I won't, you'll see when you come back :smile: Enjoy the holidays!

Now that I think about it and that we've worked with `eslint-plugin-import` a bit, I think this would be a better fit there. What do you think @benmosher?

**Comment by benmosher** [comment]
Sounds like: https://github.com/benmosher/eslint-plugin-import/issues/394 which I'm up for.

I skimmed this thread, but is what you're looking for applicable only to `.` or also `./foo`?

**Comment by jfmengels** [comment]
Hah. I actually :+1:'d that one. Totally slipped my mind.

Originally, this issue is only for the index of the current folder.

Now that you mention it, the same will be applicable to the parent (or its parent, etc.) `..` vs `../`.

I'd make this more generic and also apply it to `./foo`, to prevent `./foo/`,  `./foo/index` or `./foo/index.js`, but the index (also parent index?) is a bit special as `.` is not valid in 0.12.

**Comment by benmosher** [comment]
I'd be inclined not to support 0.12 since it's August, now. The majority of this thread is from before 0.12 phased into maintenance mode, IIRC. I could go either way, though. Would probably want the default to be 0.12-ignorant and put `./`+`../` exceptions behind a flag, if anything, just for consistency.

**Comment by sindresorhus** [comment]
No point in supporting nor having an option for 0.12 at this point. Really! Even ESLint have dropped support for it.

**Comment by jfmengels** [comment]
True. And if it does, they can disable this rule.

@sindresorhus Does that mean you'll start writing packages in Node v4 ES6 code from now on? Or soon?

**Comment by sindresorhus** [comment]
@jfmengels I've been writing new modules with Node.js 4 as target for a while now.
